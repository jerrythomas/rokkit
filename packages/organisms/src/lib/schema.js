import { typeOf } from '@rokkit/data'

/**
 * Derives a schema for properties of an object.
 *
 * @param {Object} data
 * @returns
 */
function deriveObjectProperties(data) {
	const properties = {}
	for (const [key, value] of Object.entries(data)) {
		properties[key] = deriveSchemaFromValue(value)
	}
	return properties
}

/**
 * Derives a schema from a given value.
 *
 * @param {any} data
 * @returns {import('../types').DataSchema}
 */
export function deriveSchemaFromValue(data) {
	const schema = { type: typeOf(data) }
	if (schema.type === 'array') {
		schema.items = deriveSchemaFromValue(data.length > 0 ? data[0] : {})
	} else if (schema.type === 'object') {
		schema.properties = deriveObjectProperties(data)
	}
	return schema
}
