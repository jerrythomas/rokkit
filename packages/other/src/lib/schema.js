/**
 * @typedef {Object<string, *>} VariantProperties
 */
/**
 * @typedef SharedProperties
 * @property {string} [class]
 * @property {string} [type]
 * @property {string} [label]
 * @property {string} description
 * @property {string} [placeholder]
 * @property {*} [defaultValue]
 */

/**
 * @typedef {VariantProperties & SharedProperties} Properties
 */

/**
 * @typedef AttributeSchema
 * @property {string} key
 * @property {string} [component]
 * @property {boolean} [group]
 * @property {Array<import('./rules').ValidationRule>} [rules]
 * @property {Properties} [props]
 */
