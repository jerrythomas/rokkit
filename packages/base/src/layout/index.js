export { default as Menu } from './Menu.svelte'
export { default as Sidebar } from './Sidebar.svelte'
export { default as SplitPane } from './SplitPane.svelte'
export { default as Stepper } from './Stepper.svelte'
export { default as ProgressDots } from './ProgressDots.svelte'
export { default as Icon } from './Icon.svelte'
