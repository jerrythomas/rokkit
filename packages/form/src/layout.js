import Ajv from 'ajv'
import { omit } from 'ramda'

const mapTypes = {
	enum: (props) => (props.items.length > 3 ? 'select' : 'radio'),
	uuid: () => 'text',
	float: () => 'number',
	integer: () => 'number',
	string: (props) => (props.maxLength < 255 ? 'text' : 'textarea'),
	boolean: () => 'checkbox'
}

const mapPatterns = {
	float: () => '^[0-9]+.[0-9]+$',
	integer: () => '^[0-9]+$',
	uuid: () =>
		'^[0-9a-f]{8}-[0-9a-f]{4}-[0-5][0-9a-f]{3}-[089ab][0-9a-f]{3}-[0-9a-f]{12}$'
}

export function validate(schema, data) {
	const ajv = new Ajv({ allErrors: true })
	ajv.addVocabulary(['hasChanged', 'widget'])
	const valid = ajv.validate(schema, data)
	let result = { ...schema }

	if (!valid) {
		ajv.errors.map((err) => {
			let ref = result
			err.schemaPath
				.split('/')
				.slice(1, -1)
				.map((k) => {
					ref[k] = { ...ref[k] }
					ref = ref[k]
				})
			if (ref['hasChanged']) ref.error = err.message
		})
	}
	return result
}

export function deriveLayout(schema) {
	return [schema]
}

export function propsFromSchema(props) {
	let { component, type, pattern } = props
	let result = { ...omit(['component'], props) }
	if (!component) {
		component = 'Input'
	}
	if (type in mapTypes) {
		result = { ...result, type: mapTypes[type](props) }
	}

	if (type in mapPatterns && !pattern) {
		result = { pattern: mapPatterns[type](), ...result }
	}
	return { component, properties: result }
}
