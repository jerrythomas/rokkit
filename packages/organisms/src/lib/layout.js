import { isObject } from '@rokkit/core'

/**
 * Derives a layout from a given value.
 * @param {any} value
 * @param {string} scope
 * @returns {import('../types').DataLayout}
 */
export function deriveLayoutFromValue(value, scope = '#') {
	if (Array.isArray(value)) {
		return deriveArrayLayout(value, scope)
	} else if (typeof value === 'object' && value !== null) {
		return deriveObjectLayout(value, scope)
	}
	return { type: 'vertical', elements: [{ scope }] }
}

/**
 * Derives a layout from a given array value.
 * @param {Array<any>} value
 * @param {string} scope
 * @returns {import('../types').DataLayout}
 */
function deriveArrayLayout(value, scope) {
	return {
		scope,
		schema: deriveLayoutFromValue(value[0], '#')
	}
}

/**
 * Derives a layout from a given object value.
 * @param {Object} value
 * @param {string} scope
 * @returns {import('../types').DataLayout}
 */
function deriveObjectLayout(value, scope) {
	const elements = Object.entries(value).map(([label, val]) =>
		deriveElementLayout(val, scope, label)
	)
	return { type: 'vertical', elements }
}

/**
 * Derives a layout from a given object value.
 * @param {Object} val
 * @param {string} scope
 * @param {string} label
 * @returns {import('../types').DataLayout}
 */
function deriveElementLayout(val, scope, label) {
	const path = `${scope}/${label}`
	if (isObject(val)) {
		return {
			title: label,
			scope: path,
			...deriveLayoutFromValue(val, path)
		}
	}
	return { label, scope: path }
}
