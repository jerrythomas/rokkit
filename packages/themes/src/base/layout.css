/* container {
	@apply overflow-hidden;
} */
container > segment {
	@apply overflow-hidden;
}
.three-col {
	@apply grid w-full h-full overflow-hidden;
}

.three-col.md {
	grid-template-columns: repeat(2, 1fr);
	grid-template-rows: repeat(2, 1fr);
	/* @apply grid-cols-2 grid-rows-2; */
	grid-template-areas:
		'c-1 c-2'
		'c-1 c-3';
}
.three-col.md > .col-1,
.three-col.lg > .col-1 {
	grid-area: c-1;
}
.three-col.md > .col-2,
.three-col.lg > .col-2 {
	grid-area: c-2;
}
.three-col.md > .col-3,
.three-col.lg > .col-3 {
	grid-area: c-3;
}
.three-col.lg {
	/* @apply grid-cols-5; */
	grid-template-columns: repeat(5, 1fr);
	grid-template-rows: repeat(2, 1fr);
	grid-template-areas:
		'c-1 c-1 c-2 c-2 c-2'
		'c-1 c-1 c-3 c-3 c-3';
}
.three-col.sm {
	@apply grid-cols-1 grid-rows-1;
}
