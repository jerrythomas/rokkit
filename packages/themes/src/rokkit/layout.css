/* Styles for ' rk-switch' */
[data-style='rokkit'] rk-field rk-switch:focus-within {
  @apply outline-none;
}
[data-style='rokkit'] rk-field rk-switch {
  @apply min-h-8 items-center justify-center rounded-sm border-none px-1;
}
[data-style='rokkit'] rk-field rk-switch rk-item {
  @apply rounded-sm leading-loose;
}
[data-style='rokkit'] rk-field rk-switch rk-item p {
  @apply justify-center;
}

[data-style='rokkit'] .horizontal > rk-field radio-group {
  @apply md:flex-row md:gap-4;
}
