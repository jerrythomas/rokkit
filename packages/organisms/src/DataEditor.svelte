<script>
	import { run } from 'svelte/legacy'

	import { setContext } from 'svelte'
	import { writable } from 'svelte/store'
	import { componentTypes } from './types'
	import Wrapper from './wrappers/Wrapper.svelte'
	import { Item } from '@rokkit/molecules'
	import Tabs from './Tabs.svelte'
	import FieldLayout from './FieldLayout.svelte'
	import { createEventDispatcher } from 'svelte'

	const dispatch = createEventDispatcher()
	const registry = writable({})
	setContext('registry', registry)

	import { deriveSchemaFromValue, deriveLayoutFromValue, getSchemaWithLayout } from './lib'

	/**
	 * @typedef {Object} Props
	 * @property {any} value
	 * @property {any} [schema]
	 * @property {any} [layout]
	 * @property {any} [using]
	 */

	/** @type {Props} */
	let {
		value = $bindable(),
		schema = $bindable(null),
		layout = $bindable(null),
		using = {}
	} = $props()

	let schemaWithLayout = $derived(getSchemaWithLayout(schema, layout))

	function handle() {
		dispatch('change', value)
	}
	$effect(() => {
		registry.set({
			editors: { ...componentTypes, ...using?.editors },
			components: { default: Item, ...using?.components },
			wrappers: { default: Wrapper, ...using?.wrappers },
			navigators: { default: Tabs, ...using?.navigators }
		})
	})
	$effect(() => {
		if (!schema) schema = deriveSchemaFromValue(value)
	})
	$effect(() => {
		if (!layout) layout = deriveLayoutFromValue(value)
	})
</script>

<FieldLayout schema={schemaWithLayout} bind:value on:change={handle} />
