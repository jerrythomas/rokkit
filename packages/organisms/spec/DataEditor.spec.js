import { describe, expect, it, beforeEach } from 'vitest'
import { cleanup, render } from '@testing-library/svelte'
import DataEditor from '../src/DataEditor.svelte'
import layout from './lib/fixtures/input-layout.json'
import schema from './lib/fixtures/input-schema.json'
import CustomField from './mocks/CustomField.svelte'

describe('DataEditor.svelte', () => {
	let value = {
		name: 'John',
		age: 30,
		address: {
			street: '123 Main St',
			city: 'New York',
			state: 'NY',
			zip: 10001
		}
	}

	beforeEach(() => cleanup())

	it('should render an object editor', () => {
		const { container } = render(DataEditor, { value })
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})

	it('should render an object editor using schema', () => {
		const { container } = render(DataEditor, { value, schema })
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})

	it('should render an object editor using layout', () => {
		const { container } = render(DataEditor, {
			value,
			layout,
			using: { custom: CustomField }
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})

	it('should render an object editor using schema & layout', () => {
		const { container } = render(DataEditor, {
			value,
			schema,
			layout,
			using: { custom: CustomField }
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})
})
