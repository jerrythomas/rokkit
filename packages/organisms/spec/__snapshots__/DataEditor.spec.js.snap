// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DataEditor.svelte > should handle changes 1`] = `
<div>
  <field-layout
    aria-label=""
    class="vertical "
  >
    <input-field
      class="flex flex-col input-string "
    >
      <label
        for="name"
      >
        name
      </label>
       
      <field
        aria-label="name"
        class="w-full flex flex-row items-center"
      >
         
         
        <input
          id="name"
          name="name"
          type="text"
        />
        
        
         
      </field>
       
    </input-field>
    
    <input-field
      class="flex flex-col input-number "
    >
      <label
        for="age"
      >
        age
      </label>
       
      <field
        aria-label="age"
        class="w-full flex flex-row items-center"
      >
         
         
        <input
          id="age"
          max="100"
          min="0"
          name="age"
          type="number"
        />
        
        
         
      </field>
       
    </input-field>
    
    <field-layout
      aria-label="address"
      class="vertical "
    >
      <input-field
        class="flex flex-col input-string "
      >
        <label
          for="address.street"
        >
          street
        </label>
         
        <field
          aria-label="street"
          class="w-full flex flex-row items-center"
        >
           
           
          <input
            id="address.street"
            name="address.street"
            type="text"
          />
          
          
           
        </field>
         
      </input-field>
      
      <input-field
        class="flex flex-col input-string "
      >
        <label
          for="address.city"
        >
          city
        </label>
         
        <field
          aria-label="city"
          class="w-full flex flex-row items-center"
        >
           
           
          <input
            id="address.city"
            name="address.city"
            type="text"
          />
          
          
           
        </field>
         
      </input-field>
      
      <input-field
        class="flex flex-col input-string "
      >
        <label
          for="address.state"
        >
          state
        </label>
         
        <field
          aria-label="state"
          class="w-full flex flex-row items-center"
        >
           
           
          <input
            id="address.state"
            name="address.state"
            type="text"
          />
          
          
           
        </field>
         
      </input-field>
      
      <input-field
        class="flex flex-col input-number "
      >
        <label
          for="address.zip"
        >
          zip
        </label>
         
        <field
          aria-label="zip"
          class="w-full flex flex-row items-center"
        >
           
           
          <input
            id="address.zip"
            max="99999"
            min="0"
            name="address.zip"
            type="number"
          />
          
          
           
        </field>
         
      </input-field>
      
      
    </field-layout>
    
    
    
    
    
  </field-layout>
  
  
</div>
`;

exports[`DataEditor.svelte > should handle changes 2`] = `
<div>
  <field-layout
    aria-label=""
    class="vertical "
  >
     
    <p>
      John
    </p>
    
    
    
    <input-field
      class="flex flex-col input-number "
    >
      <label
        for="age"
      >
        age
      </label>
       
      <field
        aria-label="age"
        class="w-full flex flex-row items-center"
      >
         
         
        <input
          id="age"
          max="100"
          min="0"
          name="age"
          type="number"
        />
        
        
         
      </field>
       
    </input-field>
    
    <field-layout
      aria-label="Address"
      class="vertical "
    >
      <input-field
        class="flex flex-col input-string "
      >
        <label
          for="address.street"
        >
          street
        </label>
         
        <field
          aria-label="street"
          class="w-full flex flex-row items-center"
        >
           
           
          <input
            id="address.street"
            name="address.street"
            type="text"
          />
          
          
           
        </field>
         
      </input-field>
      
      <input-field
        class="flex flex-col input-string "
      >
        <label
          for="address.city"
        >
          city
        </label>
         
        <field
          aria-label="city"
          class="w-full flex flex-row items-center"
        >
           
           
          <input
            id="address.city"
            name="address.city"
            type="text"
          />
          
          
           
        </field>
         
      </input-field>
      
      <input-field
        class="flex flex-col input-string "
      >
        <label
          for="address.state"
        >
          state
        </label>
         
        <field
          aria-label="state"
          class="w-full flex flex-row items-center"
        >
           
           
          <input
            id="address.state"
            name="address.state"
            type="text"
          />
          
          
           
        </field>
         
      </input-field>
      
      <input-field
        class="flex flex-col input-number "
      >
        <label
          for="address.zip"
        >
          zip
        </label>
         
        <field
          aria-label="zip"
          class="w-full flex flex-row items-center"
        >
           
           
          <input
            id="address.zip"
            max="99999"
            min="0"
            name="address.zip"
            type="number"
          />
          
          
           
        </field>
         
      </input-field>
      
      
    </field-layout>
    
    
    
    
    
  </field-layout>
  
  
</div>
`;

exports[`DataEditor.svelte > should render an object editor 1`] = `
<div>
  <field-layout
    aria-label=""
    class="vertical "
  >
    <input-field
      class="flex flex-col input-string "
    >
      <label
        for="name"
      >
        name
      </label>
       
      <field
        aria-label="name"
        class="w-full flex flex-row items-center"
      >
         
         
        <input
          id="name"
          name="name"
          type="text"
        />
        
        
         
      </field>
       
    </input-field>
    
    <input-field
      class="flex flex-col input-number "
    >
      <label
        for="age"
      >
        age
      </label>
       
      <field
        aria-label="age"
        class="w-full flex flex-row items-center"
      >
         
         
        <input
          id="age"
          name="age"
          type="number"
        />
        
        
         
      </field>
       
    </input-field>
    
    <field-layout
      aria-label="address"
      class="vertical "
    >
      <input-field
        class="flex flex-col input-string "
      >
        <label
          for="address.street"
        >
          street
        </label>
         
        <field
          aria-label="street"
          class="w-full flex flex-row items-center"
        >
           
           
          <input
            id="address.street"
            name="address.street"
            type="text"
          />
          
          
           
        </field>
         
      </input-field>
      
      <input-field
        class="flex flex-col input-string "
      >
        <label
          for="address.city"
        >
          city
        </label>
         
        <field
          aria-label="city"
          class="w-full flex flex-row items-center"
        >
           
           
          <input
            id="address.city"
            name="address.city"
            type="text"
          />
          
          
           
        </field>
         
      </input-field>
      
      <input-field
        class="flex flex-col input-string "
      >
        <label
          for="address.state"
        >
          state
        </label>
         
        <field
          aria-label="state"
          class="w-full flex flex-row items-center"
        >
           
           
          <input
            id="address.state"
            name="address.state"
            type="text"
          />
          
          
           
        </field>
         
      </input-field>
      
      <input-field
        class="flex flex-col input-number "
      >
        <label
          for="address.zip"
        >
          zip
        </label>
         
        <field
          aria-label="zip"
          class="w-full flex flex-row items-center"
        >
           
           
          <input
            id="address.zip"
            name="address.zip"
            type="number"
          />
          
          
           
        </field>
         
      </input-field>
      
      
    </field-layout>
    
    
    
    
    
  </field-layout>
  
  
</div>
`;

exports[`DataEditor.svelte > should render an object editor using layout 1`] = `
<div>
  <field-layout
    aria-label=""
    class="vertical "
  >
     
    <p>
      John
    </p>
    
    
    
    <input-field
      class="flex flex-col input-number "
    >
      <label
        for="age"
      >
        age
      </label>
       
      <field
        aria-label="age"
        class="w-full flex flex-row items-center"
      >
         
         
        <input
          id="age"
          name="age"
          type="number"
        />
        
        
         
      </field>
       
    </input-field>
    
    <field-layout
      aria-label="Address"
      class="vertical "
    >
      <input-field
        class="flex flex-col input-string "
      >
        <label
          for="address.street"
        >
          street
        </label>
         
        <field
          aria-label="street"
          class="w-full flex flex-row items-center"
        >
           
           
          <input
            id="address.street"
            name="address.street"
            type="text"
          />
          
          
           
        </field>
         
      </input-field>
      
      <input-field
        class="flex flex-col input-string "
      >
        <label
          for="address.city"
        >
          city
        </label>
         
        <field
          aria-label="city"
          class="w-full flex flex-row items-center"
        >
           
           
          <input
            id="address.city"
            name="address.city"
            type="text"
          />
          
          
           
        </field>
         
      </input-field>
      
      <input-field
        class="flex flex-col input-string "
      >
        <label
          for="address.state"
        >
          state
        </label>
         
        <field
          aria-label="state"
          class="w-full flex flex-row items-center"
        >
           
           
          <input
            id="address.state"
            name="address.state"
            type="text"
          />
          
          
           
        </field>
         
      </input-field>
      
      <input-field
        class="flex flex-col input-number "
      >
        <label
          for="address.zip"
        >
          zip
        </label>
         
        <field
          aria-label="zip"
          class="w-full flex flex-row items-center"
        >
           
           
          <input
            id="address.zip"
            name="address.zip"
            type="number"
          />
          
          
           
        </field>
         
      </input-field>
      
      
    </field-layout>
    
    
    
    
    
  </field-layout>
  
  
</div>
`;

exports[`DataEditor.svelte > should render an object editor using schema & layout 1`] = `
<div>
  <field-layout
    aria-label=""
    class="vertical "
  >
     
    <p>
      John
    </p>
    
    
    
    <input-field
      class="flex flex-col input-number "
    >
      <label
        for="age"
      >
        age
      </label>
       
      <field
        aria-label="age"
        class="w-full flex flex-row items-center"
      >
         
         
        <input
          id="age"
          max="100"
          min="0"
          name="age"
          type="number"
        />
        
        
         
      </field>
       
    </input-field>
    
    <field-layout
      aria-label="Address"
      class="vertical "
    >
      <input-field
        class="flex flex-col input-string "
      >
        <label
          for="address.street"
        >
          street
        </label>
         
        <field
          aria-label="street"
          class="w-full flex flex-row items-center"
        >
           
           
          <input
            id="address.street"
            name="address.street"
            type="text"
          />
          
          
           
        </field>
         
      </input-field>
      
      <input-field
        class="flex flex-col input-string "
      >
        <label
          for="address.city"
        >
          city
        </label>
         
        <field
          aria-label="city"
          class="w-full flex flex-row items-center"
        >
           
           
          <input
            id="address.city"
            name="address.city"
            type="text"
          />
          
          
           
        </field>
         
      </input-field>
      
      <input-field
        class="flex flex-col input-string "
      >
        <label
          for="address.state"
        >
          state
        </label>
         
        <field
          aria-label="state"
          class="w-full flex flex-row items-center"
        >
           
           
          <input
            id="address.state"
            name="address.state"
            type="text"
          />
          
          
           
        </field>
         
      </input-field>
      
      <input-field
        class="flex flex-col input-number "
      >
        <label
          for="address.zip"
        >
          zip
        </label>
         
        <field
          aria-label="zip"
          class="w-full flex flex-row items-center"
        >
           
           
          <input
            id="address.zip"
            max="99999"
            min="0"
            name="address.zip"
            type="number"
          />
          
          
           
        </field>
         
      </input-field>
      
      
    </field-layout>
    
    
    
    
    
  </field-layout>
  
  
</div>
`;

exports[`DataEditor.svelte > should render an object editor using schema 1`] = `
<div>
  <field-layout
    aria-label=""
    class="vertical "
  >
    <input-field
      class="flex flex-col input-string "
    >
      <label
        for="name"
      >
        name
      </label>
       
      <field
        aria-label="name"
        class="w-full flex flex-row items-center"
      >
         
         
        <input
          id="name"
          name="name"
          type="text"
        />
        
        
         
      </field>
       
    </input-field>
    
    <input-field
      class="flex flex-col input-number "
    >
      <label
        for="age"
      >
        age
      </label>
       
      <field
        aria-label="age"
        class="w-full flex flex-row items-center"
      >
         
         
        <input
          id="age"
          max="100"
          min="0"
          name="age"
          type="number"
        />
        
        
         
      </field>
       
    </input-field>
    
    <field-layout
      aria-label="address"
      class="vertical "
    >
      <input-field
        class="flex flex-col input-string "
      >
        <label
          for="address.street"
        >
          street
        </label>
         
        <field
          aria-label="street"
          class="w-full flex flex-row items-center"
        >
           
           
          <input
            id="address.street"
            name="address.street"
            type="text"
          />
          
          
           
        </field>
         
      </input-field>
      
      <input-field
        class="flex flex-col input-string "
      >
        <label
          for="address.city"
        >
          city
        </label>
         
        <field
          aria-label="city"
          class="w-full flex flex-row items-center"
        >
           
           
          <input
            id="address.city"
            name="address.city"
            type="text"
          />
          
          
           
        </field>
         
      </input-field>
      
      <input-field
        class="flex flex-col input-string "
      >
        <label
          for="address.state"
        >
          state
        </label>
         
        <field
          aria-label="state"
          class="w-full flex flex-row items-center"
        >
           
           
          <input
            id="address.state"
            name="address.state"
            type="text"
          />
          
          
           
        </field>
         
      </input-field>
      
      <input-field
        class="flex flex-col input-number "
      >
        <label
          for="address.zip"
        >
          zip
        </label>
         
        <field
          aria-label="zip"
          class="w-full flex flex-row items-center"
        >
           
           
          <input
            id="address.zip"
            max="99999"
            min="0"
            name="address.zip"
            type="number"
          />
          
          
           
        </field>
         
      </input-field>
      
      
    </field-layout>
    
    
    
    
    
  </field-layout>
  
  
</div>
`;
